### ISSUE TITLE
[Dynamic Shape] Input with dynamic shape.

### ISSUE THREAD

**Issue by baoleai** [pytorch/xla#6413]
## 🚀 Feature

Supports input with dynamic shape.

## Motivation
During the training of Large Language Models (LLMs), the sequence lengths of input data are typically variable, necessitating padding prior to training. Common padding techniques include "longest," which pads to the length of the longest sequence within a batch, and "max_length," which pads to a predetermined maximum length. Using max_length padding introduces substantial inefficiencies as the padding contributes to the computations, adversely affecting overall performance. Meanwhile, employing the longest sequence padding, although it considerably reduces redundant computations, can trigger XLA to perform multiple compilations due to the dynamic nature of the input data, such as the varying sequence_length dimension in a two-dimensional [batch_size, sequence_length] input. This dynamic shape can slow down the overall training process. While PyTorch/XLA has incorporated some dynamic operations like `masked_select` and `nonzero`, it does not yet accommodate inputs with dynamic shapes. Consequently, enabling support for dynamic shape inputs is essential.

**Comment by baoleai** [comment]
Based on https://github.com/huggingface/transformers/blob/main/src/transformers/trainer_pt_utils.py#L480, it seems that the dynamic shape feature is not supported in the GPU environment on the current master branch.

**Comment by baoleai** [comment]
During the training of LLMs such as the Llama model, the input is typically a 2D array with dimensions [batch_size, sequence_length], where the batch_size is fixed, and the sequence_length is variable. We first pad the sequences within the batch to match the length of the longest sequence, and then we proceed with training. As a result, the shape of different input batches can vary, which can be represented dynamically as [batch_size, <=max_length]. This input then sequentially passes through the embedding layer (outputting [batch_size, <=max_length, hidden_dim]), the attention layer (outputting [batch_size, <=max_length, hidden_dim]), and the MLP layer (outputting [batch_size, <=max_length, hidden_dim]) for computation. Finally, an output of shape [batch_size, <=max_length] is produced, which, along with the label [batch_size, <=max_length], is used to calculate the loss.

To illustrate the problem we're currently facing, let's start with a very simple MLP model. Once the issue is resolved with the MLP, I will retest on the actual Llama model.

I constructed the test based on https://github.com/pytorch/xla/blob/master/test/ds/test_dynamic_shape_models.py, with the main modification being to the `create_dynamic_test_data` function to set up the test.
```
import argparse
import os
import sys

parser = argparse.ArgumentParser(add_help=False)
parser.add_argument('--verbosity', type=int, default=2)
FLAGS, leftovers = parser.parse_known_args()
sys.argv = [sys.argv[0]] + leftovers

import numpy as np
import unittest
import torch
import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met

# It enables us to run python implementations of CompositeAutogradImplicit ops.
# CompositeAutogradImplicit means we don't have an explicit backward formula for an op instead an op is composed of a bunch of ops that do have backward formulas and combines this formulas is equivalent to differentiating the op explicitly.
pd = torch._C._EnablePythonDispatcher()
xla_dev = xm.xla_device()


class Feedforward(torch.nn.Module):

  def __init__(self, input_size, hidden_size):
    super().__init__()
    self.input_size = input_size
    self.hidden_size = hidden_size
    self.fc1 = torch.nn.Linear(self.input_size, self.hidden_size)
    self.fc1.weight.data.fill_(0.01)
    self.fc1.bias.data.fill_(0.01)
    self.relu = torch.nn.ReLU()
    self.fc2 = torch.nn.Linear(self.hidden_size, 1)
    self.fc2.weight.data.fill_(0.01)
    self.fc2.bias.data.fill_(0.01)
    self.sigmoid = torch.nn.Sigmoid()

  def forward(self, x):
    hidden = self.fc1(x)
    relu = self.relu(hidden)
    output = self.fc2(relu)
    output = self.sigmoid(output)
    return output


@unittest.skipIf(
    not xm.get_xla_supported_devices("GPU") and
    not xm.get_xla_supported_devices("TPU"),
    f"The tests fail on CPU. See https://github.com/pytorch/xla/issues/4298 for more detail."
)
class TestDynamicShapeModels(unittest.TestCase):

  def test_forward_pass_dynamic_input_correctness(self):
    losses = []
    for _ in range(2):
      num_features = 2
      num_test_samples = 10
      x_test, y_test = self.create_dynamic_test_data(num_test_samples,
                                                     num_features, xla_dev)

      model = Feedforward(num_features, hidden_size=10).to(xla_dev)
      criterion = torch.nn.BCELoss()

      model.eval()
      with torch.no_grad():
        y_pred = model(x_test)
        before_train = criterion(y_pred.squeeze(), y_test)
        xm.mark_step()
        losses.append(before_train.item())

    np.testing.assert_allclose(losses[0], losses[1], rtol=1e-2, atol=1e-2)
    print('Test passed.')

  def test_forward_pass_dynamic_input_compile_once(self):
    met.clear_metrics()
    num_compilation_recorded = False
    num_compilation = -1
    for i in range(10):
      num_features = 2
      num_test_samples = 10
      x_test, y_test = self.create_dynamic_test_data(
          num_test_samples, num_features, xla_dev, num_non_zeros=i)

      model = Feedforward(num_features, hidden_size=10).to(xla_dev)
      criterion = torch.nn.BCELoss()

      model.eval()
      with torch.no_grad():
        y_pred = model(x_test)
        criterion(y_pred.squeeze(), y_test)
        xm.mark_step()
        if not num_compilation_recorded:
          num_compilation = met.metric_data('CompileTime')[0]
          num_compilation_recorded = True
        else:
          self.assertEqual(num_compilation,
                           met.metric_data('CompileTime')[0],
                           'number of compilation should not increase.')

  def test_backward_pass_with_dynamic_input(self):
    num_features = 2
    num_test_samples = 10
    model = Feedforward(num_features, hidden_size=10).to(xla_dev)
    criterion = torch.nn.BCELoss()
    optimizer = torch.optim.SGD(model.parameters(), lr=1e-3)
    optimizer.zero_grad()

    # training
    model.train()
    x_training, y_training = self.create_dynamic_test_data(
        num_test_samples, num_features, xla_dev)
    y_pred = model(x_training)
    loss = criterion(y_pred.squeeze(), y_training)
    # Backpropagation.
    loss.backward()
    xm.optimizer_step(optimizer)
    print('Finished training.')

    # testing
    model.eval()
    with torch.no_grad():
      x_test, y_test = self.create_dynamic_test_data(num_test_samples,
                                                     num_features, xla_dev)
      y_pred = model(x_test)
      criterion(y_pred.squeeze(), y_test).item()
      xm.mark_step()
    print('Test passed.')

  def create_dynamic_test_data(self,
                               num_test_samples,
                               num_features,
                               device,
                               num_non_zeros=10):
    x_test = torch.zeros(num_test_samples, num_features)
    y_test = torch.zeros(num_test_samples)
    for i in range (num_non_zeros):
      for j in range(num_features):
        x_test[i][j] = 1
      y_test[i] = 1

    x_test_xla = x_test.to(device)
    y_test_xla = y_test.to(device)

    # It should be noted that we cannot directly reshape the output of masked_select into [<=10, 2].
    # Therefore, we apply masked_select to each sample in the batch individually and subsequently
    # stack them to form a 2D tensor.
    x_test_nonzero_dev1 = torch.masked_select(x_test_xla[:, 0].int(), x_test_xla[:, 0] != 0)
    x_test_nonzero_dev2 = torch.masked_select(x_test_xla[:, 1].int(), x_test_xla[:, 1] != 0)

    # x_test_nonzero_dev is with shape [<=num_test_samples, 2]
    # y_test_nonzero_dev is with shape [<=num_test_samples]
    x_test_nonzero_dev = torch.stack([x_test_nonzero_dev1, x_test_nonzero_dev2], dim=1).float()
    y_test_nonzero_dev = torch.masked_select(y_test_xla.int(), y_test_xla != 0).float().squeeze()

    # uncommenting the line below will allow the test to pass.
    # print(x_test_nonzero_dev, y_test_nonzero_dev)

    return x_test_nonzero_dev, y_test_nonzero_dev


if __name__ == '__main__':
  assert os.environ['XLA_EXPERIMENTAL'] != ''
  test = unittest.main(verbosity=FLAGS.verbosity, exit=False)
  # DISABLE PYTHON DISPATCHER FLAG
  del pd
  sys.exit(0 if test.result.wasSuccessful() else 1)
```

The input is shaped as [<=10, 2], and the label shape is [<=10] to simulate the variation of batch_size * sequence_length.

The test environment is as follows:

- pytorch/xla version v2.1.0
- GPU A100

The test script is run as follows:

```
PJRT_DEVICE=GPU XLA_EXPERIMENTAL=nonzero:masked_select python pytorch/xla/test/ds/test_dynamic_shape_models.py -v
```

It was found that when I print the `x_test_nonzero_dev` and `y_test_nonzero_dev`, the test runs successfully. Otherwise a coredump occurs.

The error message is as follows: 
```
LLVM ERROR: Expect an prototype attribute to obtain the underlying dialect but got none
```

The stack of coredump is as follows:
```
#1  0x00007f1399d6f7f1 in __GI_abort () at abort.c:79
#2  0x00007f1300a6f90f in llvm::report_fatal_error(llvm::Twine const&, bool) [clone .cold] () from /root/pytorch/xla/_XLAC.cpython-38-x86_64-linux-gnu.so
#3  0x00007f1309e9e8cb in llvm::report_fatal_error(char const*, bool) () from /root/pytorch/xla/_XLAC.cpython-38-x86_64-linux-gnu.so
#4  0x00007f130927c408 in mlir::hlo::boundsToEncoding(mlir::Attribute, llvm::ArrayRef<long>) () from /root/pytorch/xla/_XLAC.cpython-38-x86_64-linux-gnu.so
#5  0x00007f13091c1ec5 in mlir::hlo::inferConcatenateOp(std::optional<mlir::Location>, mlir::TypeRange, long, llvm::SmallVectorImpl<mlir::Type>&) ()
   from /root/pytorch/xla/_XLAC.cpython-38-x86_64-linux-gnu.so
#6  0x00007f1307bb0480 in mlir::mhlo::ConcatenateOp::inferReturnTypes(mlir::MLIRContext*, std::optional<mlir::Location>, mlir::ValueRange, mlir::DictionaryAttr, mlir::OpaqueProperties, mlir::RegionRange, llvm::SmallVectorImpl<mlir::Type>&) () from /root/pytorch/xla/_XLAC.cpython-38-x86_64-linux-gnu.so
#7  0x00007f1307bb0b1b in mlir::detail::InferTypeOpInterfaceInterfaceTraits::Model<mlir::mhlo::ConcatenateOp>::refineReturnTypes(mlir::MLIRContext*, std::optional<mlir::Location>, mlir::ValueRange, mlir::DictionaryAttr, mlir::OpaqueProperties, mlir::RegionRange, llvm::SmallVectorImpl<mlir::Type>&) ()
   from /root/pytorch/xla/_XLAC.cpython-38-x86_64-linux-gnu.so
#8  0x00007f1309ca5764 in mlir::detail::verifyInferredResultTypes(mlir::Operation*) () from /root/pytorch/xla/_XLAC.cpython-38-x86_64-linux-gnu.so
#9  0x00007f1307aca3f9 in mlir::Op<mlir::mhlo::ConcatenateOp, mlir::OpTrait::ZeroRegions, mlir::OpTrait::OneResult, mlir::OpTrait::OneTypedResult<mlir::TensorType>::Impl, mlir::OpTrait::ZeroSuccessors, mlir::OpTrait::VariadicOperands, mlir::OpTrait::OpInvariants, mlir::ConditionallySpeculatable::Trait, mlir::OpTrait::AlwaysSpeculatableImplTrait, mlir::MemoryEffectOpInterface::Trait, mlir::OpTrait::SameOperandsAndResultElementType, mlir::InferTypeOpInterface::Trait, mlir::InferShapedTypeOpInterface::Trait>::verifyRegionInvariants(mlir::Operation*) ()
   from /root/pytorch/xla/_XLAC.cpython-38-x86_64-linux-gnu.so
#10 0x00007f1307c1b39f in mlir::RegisteredOperationName::Model<mlir::mhlo::ConcatenateOp>::verifyRegionInvariants(mlir::Operation*) ()
   from /root/pytorch/xla/_XLAC.cpython-38-x86_64-linux-gnu.so
#11 0x00007f1309e0a6b0 in (anonymous namespace)::OperationVerifier::verifyOperation(mlir::Operation&) () from /root/pytorch/xla/_XLAC.cpython-38-x86_64-linux-gnu.so
#12 0x00007f1309e0be79 in (anonymous namespace)::OperationVerifier::verifyOpAndDominance(mlir::Operation&) ()
```

**Comment by JackCaoG** [comment]
`We first pad the sequences within the batch to match the length of the longest sequence, and then we proceed with training. As a result, the shape of different input batches can vary`

Can you pad to a fixed longest sequence instead of the longest sequence in the batch? And yea bounded dynamic shape is a WIP project. We paused the development on second half of 2023 but we are considering picking it up again. Would you guys be interested on contributing to the development?

**Comment by baoleai** [comment]
We are currently using a fixed max sequence, but this approach leads to the padding parts also being involved in computation and memory allocation, resulting in a waste of computation and memory, which in turn affects overall throughput. Therefore, we want to try padding according to the length of the longest sequence within the batch, hoping to reduce some unnecessary computations. I want to confirm whether bounded dynamic shape can achieve this purpose, or if it only solves the recompilation issue. If it cannot save on computation, then bounded dynamic shape might not meet my needs.

**Comment by JackCaoG** [comment]
bounded dynamic shape will not save compute for you, you can think of it as equivalent of you manually padding to a fixed length. BTW I think we(@Liyang90) also implemented the thing you guys described for our llama model. Pretty much pick a couple different max_seq, compile the graph and pick the smallest one that fit during the runtime.

**Comment by baoleai** [comment]
Thanks Jack. We have also employed some workaround methods. For example, we divide the sequence lengths into several buckets, and then for each batch, we choose the nearest bucket to determine the size of the padding. This method can indeed reduce invalid computations to a certain extent. However, the time cost of multiple compilations also increases significantly. By the way, do you have any suggestions for accelerating compilation, such as multithreaded compilation?

**Comment by JackCaoG** [comment]
@baoleai I am not sure what's the default behavior on the GPU, but on TPU compilation will usually uses all of the cpu resource available. https://github.com/pytorch/xla/blob/master/API_GUIDE.md#compilation-caching will help you not recompiling, but it won't help if you change your model code..
